### ISSUE TITLE
TestCaseData dynamic parameters

### ISSUE THREAD

**Issue by iamgithb** [nunit/nunit#3948]
Hello, 
I have some situation when need accumulate data by switch case, append it to array
![image](https://user-images.githubusercontent.com/90416477/135523230-1be47865-f7a6-427b-826c-bb2c58f0a049.png)
and transfer case data array to the test 
![image](https://user-images.githubusercontent.com/90416477/135523274-d874bcd8-1460-4016-96fe-a2271fd6584e.png)
I use object[] args and ofc i see error descripted that you needed put to much arguments.

I understand that is not in the specification.
I cant use external object how to argument
![image](https://user-images.githubusercontent.com/90416477/135525016-7f9466da-afdd-4256-a6ab-c89abfb85e66.png)

How i can work dynamical with parameters? 
Should I write the code differently?

**Comment by jnm2** [comment]
Yes, you need to write the code slightly differently due to the limitations of .NET attribute metadata. I think https://docs.nunit.org/articles/nunit/writing-tests/attributes/testcasesource.html is what you would want.

**Comment by mikkelbu** [comment]
@iamgithb I'm sorry, but I'm not sure I understand exactly what you want to do. Can you provide a more complete example?

Please note that it is not a requirement from NUnit but a general requirement from C# that attributes can only have arguments that are constants. From where should `parametrs` come?

**Comment by iamgithb** [comment]
> @iamgithb I'm sorry, but I'm not sure I understand exactly what you want to do. Can you provide a more complete example?
> 
> Please note that it is not a requirement from NUnit but a general requirement from C# that attributes can only have arguments that are constants. From where should `parametrs` come?

Why is requirement c#, in c# i can do: `void func(params string[] args)`, then `func('1','1','2' ... )`
 
[TestCaseSource("function", object[] { 'arg1', 'arg2'})
And if i init: `static object[] function(params or string[] args)` it will doesnt work
I should do references all args: `static object[] function(string arg1, string arg2, string arg3) { }`
![image](https://user-images.githubusercontent.com/90416477/135536588-972d7733-8ccf-4b75-b8b8-46908c096fb2.png)

I slightly understand why it works that way. I search different way

I found, sorry. But i should be Upper this question 🙋 _important_
https://github.com/nunit/nunit/issues/1883

**Comment by jnm2** [comment]
> Why is requirement c#

The C# language specification limits what kinds of things you can use as attribute arguments. https://github.com/dotnet/csharplang/discussions is a good place to ask questions to find out more about why C# is this way.

> `[TestCaseSource("function", object[] { 'arg1', 'arg2'})]`

The idea of TestCaseSource is that you something similar to this instead:

```cs
[TestCaseSource("SomeTestCases")]
public void SomeTest(string p1, string p2)
{
    // ...
}

public static IEnumerable<TestCaseData> SomeTestCases => new[]
{
    new TestCaseData("arg1", "arg2"),
    new TestCaseData(function("arg1", "arg2")),
    // etc
};
```

**Comment by iamgithb** [comment]
> There are several variations that NUnit also understands.
Alone I see, a classic. What are the others? _Share your experiences, please_

**Comment by mikkelbu** [comment]
@iamgithb I'm still sure exactly what you want to do, can you provide a Minimal, Reproducible Example rather than fragments of code.

Note that #1883 is about a test method with a `params` argument (were the running the tests fail). In your code above it is the test case source that has a `params` argument (were the analyzers complain), so it is not guaranteed that a fix to #1883 will fix your problem.

**Comment by rprouse** [comment]
Closing as a dormant/answered question.
